### ISSUE TITLE
Allow setting JSON_BuildTests=OFF from parent CMakeLists.txt

### ISSUE THREAD

**Issue by FDUdannychen** [nlohmann/json#846]
I think a standard BUILD_TESTING option is better in this case?

Meanwhile, setting JSON_BuildTests=ON prevents the user from disabling the unit test build from parent CMakeLists.txt.

**Comment by nlohmann** [comment]
What do you propose? Replacing `JSON_BuildTests` with `BUILD_TESTING`?

**Comment by FDUdannychen** [comment]
I want to disable building tests in my project's CMakeLists.txt. I have a `thirdparty/CMakeLists.txt`:
```
-myproj
    -src
        -CMakeLists.txt (build myproj sources)
    -thirdparty
        -json
        -another_lib
        -CMakeLists.txt (I want to disable json's unit tests here)
```

**Comment by nlohmann** [comment]
OK. I'm not a CMake expert. Could you describe what needs to be done or open a PR?

**Comment by FDUdannychen** [comment]
@nlohmann Sorry I just notice you are not using CTest. So BUILD_TESTING is not an option here. I'll pass -DBuildTests=OFF in a script instead of specify it in my CMakeLists.txt as a workaround. This issue should be closed.

**Comment by nlohmann** [comment]
In fact we are using ctest...

**Comment by theodelrieu** [comment]
I've seen quite a few cmake projects using `BUILD_TESTS` as an option. I thought it was an official option, apparently not.

IIUC, we're not including the CTest module, so the `BUILD_TESTING` option is missing?

**Comment by nlohmann** [comment]
Any ideas on this?

**Comment by TinyTinni** [comment]
I am not really sure, what the original issue is.
You can set JSON_BuildTests=OFF in your CMake script before including the json CMakeLIsts with e.g. `add_subdirectory` (don't forget declaring it as a `CACHE `variable, or using `option` instead of set).

> IIUC, we're not including the CTest module, so the BUILD_TESTING option is missing?

Yes, regarding to the [CTest](https://cmake.org/cmake/help/v3.10/module/CTest.html) doc, BUILD_TESTING is included when you include the CTest module via cmake (not done by the project atm).

The `BUILD_TESTING` seems to be a variable where you can control all tests of all projects included by cmake.

I would propose:
1) include CTest
2) add test targets only, when `BUILD_TESTING` and `JSON_BuildTests` are `ON`

This approach allows disabling all tests (of all parent/subprojects) with `BUILD_TESTING` or disabling just the json tests.

The problem would be in the existence of a  single inconsistency:
When `BUILD_TESTING` is OFF, but `JSON_BuildTests` is ON. 
Some people might think, the json tests will be build.

The default value of `BUILD_TESTING` is ON, so the approach described above is backward compatible.

ATM, I see no good approach that keeps backward compatibility and/or consistency of the 2 variables, when they get replaced and/or depends on each other.

Corresponding pull request:  #885

**Comment by nlohmann** [comment]
Closed by merging #885.
